### ISSUE TITLE
After upgrade to 1.11.1 panic when creating a collection presumably under heavy io load

### ISSUE THREAD

**Issue by volkerstampa** [qdrant/qdrant#4970]
Yesterday we upgraded to 1.11.1 and tonight we saw some panics in the Qdrant logs most often when creating collections 
but also for other operations.

## Current Behavior

During the night some automated tests create and delete collections. At the same time we presumably have some heavy disk IO ongoing as a rsync process copies files from the local disk (where also the Qdrant docker volume is kept) to NFS. Many of the create collection operations failed with:

```
2024-08-28T00:36:39.384140Z ERROR ... Error in the response: The operation was cancelled
```

At the same time the Qdrant logs show:
```
2024-08-28T00:36:38.368827Z  INFO storage::content_manager::toc::collection_meta_ops: Creating collection ...    
2024-08-28T00:36:39.019474Z ERROR qdrant::startup: Panic backtrace: 
   0: qdrant::startup::setup_panic_hook::{{closure}}
   1: std::panicking::rust_panic_with_hook
   2: std::panicking::begin_panic_handler::{{closure}}
   3: std::sys_common::backtrace::__rust_end_short_backtrace
   4: rust_begin_unwind
   5: core::panicking::panic_fmt
   6: segment::segment::entry::<impl segment::entry::entry_point::SegmentEntry for segment::segment::Segment>::info
   7: collection::shards::local_shard::shard_ops::<impl collection::shards::shard_trait::ShardOperation for collection::shards::local_shard::LocalShard>::info::{{closure}}
   8: collection::shards::replica_set::read_ops::<impl collection::shards::replica_set::ShardReplicaSet>::info::{{closure}}::{{closure}}::{{closure}}
   9: <futures_util::stream::futures_unordered::FuturesUnordered<Fut> as futures_core::stream::Stream>::poll_next
  10: collection::collection::collection_ops::<impl collection::collection::Collection>::info::{{closure}}
  11: qdrant::common::collections::do_get_collection::{{closure}}
  12: qdrant::tonic::api::collections_common::get::{{closure}}
  13: <qdrant::tonic::api::collections_api::CollectionsService as api::grpc::qdrant::collections_server::Collections>::get::{{closure}}
  14: <<api::grpc::qdrant::collections_server::CollectionsServer<T> as tower_service::Service<http::request::Request<B>>>::call::GetSvc<T> as tonic::server::service::UnaryService<api::grpc::qdrant::GetCollectionInfoRequest>>::call::{{closure}}
  15: <api::grpc::qdrant::collections_server::CollectionsServer<T> as tower_service::Service<http::request::Request<B>>>::call::{{closure}}
  16: <tower::util::map_response::MapResponseFuture<F,N> as core::future::future::Future>::poll
  17: <tonic::transport::service::router::RoutesFuture as core::future::future::Future>::poll
  18: <qdrant::tonic::tonic_telemetry::TonicTelemetryService<S> as tower_service::Service<http::request::Request<hyper::body::body::Body>>>::call::{{closure}}
  19: <qdrant::tonic::logging::LoggingMiddleware<S> as tower_service::Service<http::request::Request<hyper::body::body::Body>>>::call::{{closure}}
  20: <tonic::transport::server::SvcFuture<F> as core::future::future::Future>::poll
  21: <hyper::proto::h2::server::H2Stream<F,B> as core::future::future::Future>::poll
  22: tokio::runtime::task::raw::poll
  23: tokio::runtime::scheduler::multi_thread::worker::Context::run_task
  24: tokio::runtime::scheduler::multi_thread::worker::run
  25: tokio::runtime::task::raw::poll
  26: std::sys_common::backtrace::__rust_begin_short_backtrace
  27: core::ops::function::FnOnce::call_once{{vtable.shim}}
  28: std::sys::pal::unix::thread::Thread::new::thread_start
  29: <unknown>
  30: clone
    
2024-08-28T00:36:39.019501Z ERROR qdrant::startup: Panic occurred in file /usr/local/cargo/registry/src/index.crates.io-6f17d22bba15001f/atomic_refcell-0.1.13/src/lib.rs at line 129: already mutably borrowed    
2024-08-28T00:36:39.069768Z ERROR qdrant::startup: Panic backtrace:
   ... (same backtrace as above) ...
```

## Steps to Reproduce

Presumable this can be reproduced by creating collections while concurrently doing some heavy disk IO, however this has not been verified in isolation yet.

## Expected Behavior

I expect that collection can be created and there is no panic in Qdrant.

## Context (Environment)

This seems to occur only under special circumstances (i.e. heavy disk IO). We can probably try to avoid this in the first place.
Nontheless it might be worth investigating.

**Comment by volkerstampa** [comment]
Yes, indeed. Each test creates a "temporary" collection in the beginning then maybe inserts a few vectors and executes a search and then deletes the temporary collection at the end of the test.

**Comment by timvisee** [comment]
We have identified the problem and will be releasing a patch version shortly (with <https://github.com/qdrant/qdrant/pull/4973>).

It first appeared in Qdrant 1.11.1 so you may fall back to version 1.11.0 in the meanwhile.

I'll keep you posted on the release.

**Comment by agourlay** [comment]
Fixed in [1.11.2](https://github.com/qdrant/qdrant/releases/tag/v1.11.2)

**Comment by volkerstampa** [comment]
Thanks for the quick fix.
